### DISCUSSION TITLE
Receiving "Check failed: LayoutUtil::IsMonotonicWithDim0Major" error after upgrading from 0.4.29 to >=0.4.30

### DISCUSSION THREAD

**Discussion by Chulabhaya** [jax-ml/jax#23859] [Q&A] [answered]
Hi there! My code that was functioning at Jax versions <=0.4.29 is now failing after upgrading to any version beyond 0.4.30. In particular, my code now crashes after a short period with the following:

```
F0922 22:52:09.660401 1439192 reduction.cc:147] Check failed: LayoutUtil::IsMonotonicWithDim0Major( hero.instruction().operand(i)->shape().layout()) reduction-layout-normalizer must run before code generation
*** Check failure stack trace: ***
    @     0x7f669b0001d4  absl::lts_20230802::log_internal::LogMessage::SendToLog()
    @     0x7f669b0000d4  absl::lts_20230802::log_internal::LogMessage::Flush()
    @     0x7f669b000579  absl::lts_20230802::log_internal::LogMessageFatal::~LogMessageFatal()
    @     0x7f6696e8e13e  xla::gpu::(anonymous namespace)::ReductionEmitter::ReductionEmitter()
    @     0x7f6696e8d307  xla::gpu::ReductionFusion::EmitInitializers()
    @     0x7f6696fb1dfa  xla::gpu::KernelFusionEmitterBase::Emit()
    @     0x7f6693a704a4  xla::gpu::IrEmitterUnnested::EmitFusion()
    @     0x7f6693a7b135  xla::gpu::IrEmitterUnnested::EmitHloInstruction()
    @     0x7f6693a5dcbe  xla::gpu::IrEmitterUnnested::EmitHloComputation()
    @     0x7f6693a613f4  xla::gpu::IrEmitterUnnested::EmitCommandBufferThunk()
    @     0x7f6693a7b42c  xla::gpu::IrEmitterUnnested::EmitHloInstruction()
    @     0x7f6693a5dcbe  xla::gpu::IrEmitterUnnested::EmitHloComputation()
    @     0x7f6693a72653  xla::gpu::IrEmitterUnnested::BuildWhileThunk()
    @     0x7f6693a722b6  xla::gpu::IrEmitterUnnested::EmitWhile()
    @     0x7f6693a7b0b6  xla::gpu::IrEmitterUnnested::EmitHloInstruction()
    @     0x7f6693a5dcbe  xla::gpu::IrEmitterUnnested::EmitHloComputation()
    @     0x7f6693a5d566  xla::gpu::IrEmitterUnnested::EmitConditional()
    @     0x7f6693a7afbf  xla::gpu::IrEmitterUnnested::EmitHloInstruction()
    @     0x7f6693a5dcbe  xla::gpu::IrEmitterUnnested::EmitHloComputation()
    @     0x7f6693a72653  xla::gpu::IrEmitterUnnested::BuildWhileThunk()
    @     0x7f6693a722b6  xla::gpu::IrEmitterUnnested::EmitWhile()
    @     0x7f6693a7b0b6  xla::gpu::IrEmitterUnnested::EmitHloInstruction()
    @     0x7f6693a5dcbe  xla::gpu::IrEmitterUnnested::EmitHloComputation()
    @     0x7f6693838d2a  xla::gpu::CompileModuleToLlvmIr()
    @     0x7f6693819920  xla::gpu::GpuCompiler::CompileToBackendResult()
    @     0x7f669381c4fe  xla::gpu::GpuCompiler::RunBackend()
    @     0x7f66937dd982  xla::Service::BuildExecutable()
    @     0x7f66937a5f55  xla::LocalService::CompileExecutables()
    @     0x7f6693799a14  xla::LocalClient::Compile()
    @     0x7f669373edcb  xla::PjRtStreamExecutorClient::CompileInternal()
    @     0x7f669373fe7e  xla::PjRtStreamExecutorClient::Compile()
    @     0x7f669369766a  std::__detail::__variant::__gen_vtable_impl<>::__visit_invoke()
    @     0x7f66936872d4  pjrt::PJRT_Client_Compile()
    @     0x7f66a74fd4dd  xla::InitializeArgsAndCompile()
    @     0x7f66a74fd9be  xla::PjRtCApiClient::Compile()
    @     0x7f66ad0a016c  xla::ifrt::PjRtLoadedExecutable::Create()
    @     0x7f66ad09b5a1  xla::ifrt::PjRtCompiler::Compile()
    @     0x7f66ac8439fe  xla::PyClient::CompileIfrtProgram()
    @     0x7f66ac8448de  xla::PyClient::Compile()
    @     0x7f66ac84cbef  nanobind::detail::func_create<>()::{lambda()#1}::__invoke()
    @     0x7f66ad079631  nanobind::detail::nb_func_vectorcall_complex()
    @     0x7f6776f7d478  nanobind::detail::nb_bound_method_vectorcall()
    @     0x5605f5266fbf  PyObject_Vectorcall
Aborted
```

Would anyone happen to know what would cause this error when upgrading from 0.4.29 to a newer version of Jax?

**Comment by yashk2810** [comment]
What was your code? Do you have a minimal repro? Is this on CPU/GPU/TPU/something else? Maybe try with jax 0.4.33?

**Reply by Chulabhaya** [reply]
My code is running this file in this repo: https://github.com/FLAIROx/JaxMARL/blob/main/baselines/QLearning/qmix_rnn.py

I don't know how to make a minimal reproduction in this case because I don't know when the error happens. This is with GPU, and I have tried with all the latest Jax versions, as mentioned in my title.

The packages in my conda environment are as follows:

```
absl-py                  2.1.0
antlr4-python3-runtime   4.9.3
appdirs                  1.4.4
blinker                  1.8.1
brax                     0.10.3
certifi                  2024.8.30
charset-normalizer       3.3.2
chex                     0.1.86
click                    8.1.7
cloudpickle              3.0.0
contextlib2              21.6.0
contourpy                1.2.1
cycler                   0.12.1
decorator                5.1.1
distrax                  0.1.5
dm-env                   1.6
dm-tree                  0.1.8
docker-pycreds           0.4.0
docstring-parser         0.15
dotmap                   1.3.30
etils                    1.7.0
evosax                   0.1.5
Farama-Notifications     0.0.4
filelock                 3.15.4
flashbax                 0.1.0
Flask                    3.0.3
Flask-Cors               4.0.0
flax                     0.9.0
fonttools                4.53.0
fsspec                   2024.2.0
gast                     0.5.4
gitdb                    4.0.11
GitPython                3.1.42
glfw                     2.7.0
grpcio                   1.62.2
gym                      0.26.2
gym-notices              0.0.8
gymnasium                0.29.1
gymnax                   0.0.6
h5py                     3.10.0
huggingface-hub          0.23.4
hydra-core               1.3.2
idna                     3.6
importlib_resources      6.1.2
iniconfig                2.0.0
install                  1.3.5
itsdangerous             2.2.0
jax                      0.4.33
jax-cuda12-pjrt          0.4.33
jax-cuda12-plugin        0.4.33
jaxlib                   0.4.33
jaxmarl                  0.0.5
jaxopt                   0.8.3
Jinja2                   3.1.3
kiwisolver               1.4.5
markdown-it-py           3.0.0
MarkupSafe               2.1.5
matplotlib               3.9.0
mdurl                    0.1.2
ml_collections           0.1.1
ml-dtypes                0.4.0
msgpack                  1.0.8
mujoco                   3.1.3
mujoco-mjx               3.1.3
munkres                  1.1.4
nest-asyncio             1.6.0
numpy                    1.26.4
nvidia-cublas-cu12       12.3.4.1
nvidia-cuda-cupti-cu12   12.3.101
nvidia-cuda-nvcc-cu12    12.3.107
nvidia-cuda-nvrtc-cu12   12.3.107
nvidia-cuda-runtime-cu12 12.3.101
nvidia-cudnn-cu12        9.3.0.75
nvidia-cufft-cu12        11.0.12.1
nvidia-cusolver-cu12     11.5.4.101
nvidia-cusparse-cu12     12.2.0.103
nvidia-nccl-cu12         2.19.3
nvidia-nvjitlink-cu12    12.3.101
omegaconf                2.3.0
opt-einsum               3.3.0
optax                    0.2.3
orbax                    0.1.9
orbax-checkpoint         0.5.9
packaging                24.1
pandas                   2.2.2
patsy                    0.5.6
pettingzoo               1.24.3
pillow                   10.4.0
pip                      24.0
platformdirs             4.2.2
pluggy                   1.5.0
ply                      3.11
protobuf                 4.25.3
psutil                   5.9.8
pygame                   2.5.2
Pygments                 2.17.2
PyOpenGL                 3.1.7
pyparsing                3.1.2
PyQt5                    5.15.9
PyQt5-sip                12.12.2
pytest                   8.2.0
python-dateutil          2.9.0
pytinyrenderer           0.0.14
pytz                     2024.1
PyYAML                   6.0.1
requests                 2.31.0
rich                     13.7.1
ruff                     0.3.0
safetensors              0.4.2
scipy                    1.12.0
seaborn                  0.13.2
sentry-sdk               1.40.6
setproctitle             1.3.3
setuptools               70.0.0
shtab                    1.7.0
sip                      6.7.12
six                      1.16.0
smmap                    5.0.1
statsmodels              0.14.2
tensorboardX             2.6.2.2
tensorflow-probability   0.23.0
tensorstore              0.1.54
toml                     0.10.2
tomli                    2.0.1
toolz                    0.12.1
tornado                  6.4.1
tqdm                     4.66.2
trimesh                  4.3.2
typing_extensions        4.10.0
tyro                     0.7.3
tzdata                   2024.1
urllib3                  2.2.1
wandb                    0.17.8
Werkzeug                 3.0.2
wheel                    0.43.0
zipp                     3.17.0
```

**Reply by Chulabhaya** [reply]
If you might be able to help me understand what this Jax error even is, then that might help me narrow down what in the code might be suddenly incompatible with 0.4.30 and newer.

**Reply by nouiz** [reply]
The important part of the error is the first line: 
```
F0922 22:52:09.660401 1439192 reduction.cc:147] Check failed: LayoutUtil::IsMonotonicWithDim0Major( hero.instruction().operand(i)->shape().layout()) reduction-layout-normalizer must run before code generation
```
This point point to an error related to reduction and layout.
Can you repro single GPU? Can you dump the HLO, this should allow the XLA team to reproduce the error. You can do it with that env variable, just update the DIR to something useful:
`XLA_FLAGS="--xla_dump_to=${DIR} --xla_dump_hlo_as_text"`

**Reply by Chulabhaya** [reply]
Hi @nouiz , thank you for the reply! I've attached the HLO dump as a tar.gz file to this message, and this was run on a single GPU, it is consistently reproducible. 

[hlo_dump.tar.gz](https://github.com/user-attachments/files/17106526/hlo_dump.tar.gz)

**Reply by Chulabhaya** [reply]
Please let me know if that HLO data was generated correctly or if you would like me to re-run it or attach something different!

**Reply by akuegel** [reply]
From the stack trace, it looks like this is using the legacy reduction emitter. Since Jax version 0.4.32, we have replaced it with a new reduction emitter that should be able to handle non-default reduction layouts. I have run your HLO dump with XLA at head, and it seems to be working fine. However I also cannot reproduce it with the legacy emitter with the provided dump. Can you please try out with the newest Jax version, and provide a reproducer with that version if it still doesn't work? You can try the flag --xla_dump_hlo_as_proto instead of --xla_dump_hlo_as_text in case this is tooling related and the text dump doesn't contain some important piece.

**Reply by jreiffers** [reply, marked as answer]
This crash is in legacy emitters, which is strange, since they should be disabled by default now. Do you set any special XLA_FLAGS? Can you try setting `--xla_gpu_mlir_emitter_level=4`? But normally, this should be the default with recent versions of JAX, and the code from your stack trace shouldn't be used. Just to confirm: you are using an Nvidia GPU, right? Judging from your installed plugins that should be the case, but just to double check.

See https://jax.readthedocs.io/en/latest/xla_flags.html#configuring-xla-in-jax for instructions on how to configure XLA_FLAGS.

**Reply by Chulabhaya** [reply]
Hi @akuegel and @jreiffers , thank you guys for helping me investigate this. First off, here is the HLO dump as proto: 
[hlo_dump_proto.tar.gz](https://github.com/user-attachments/files/17119645/hlo_dump_proto.tar.gz)

As far as I know I'm not setting any special XLA_FLAGS, unless some package somewhere is doing it behind-the-scenes. 

Setting `--xla_gpu_mlir_emitter_level=4` actually did the trick and my code is running! So this appears to be the issue. I'm definitely using an Nvidia GPU as well. I'm running 0.4.33 currently, so I'm not sure why I need to enable this flag manually? Do I just need to set this variable manually every time now going forwards?

**Reply by akuegel** [reply]
0.4.33 should theoretically use a XLA revision that has the flag set to 4 by default. We are not sure why it does not work. Currently, this is guarded with a compilation flag of if_cuda_is_configured (as it doesn't work on ROCM yet), maybe for some reason in the jaxlib build this configuration for cuda was not done? Unfortunately I don't know the JAX side.

**Reply by Chulabhaya** [reply]
I see, thank you for the clarification. Are there people who are familiar with the JAX side who I should be reporting this issue to for further investigation? While setting the `--xla_gpu_mlir_emitter_level=4` does work, it would be nice to not have to do that in my code going forwards with newer versions of JAX.

**Reply by akuegel** [reply]
We have a plan to eventually remove the legacy emitters (and the flag). When exactly I cannot tell yet, but at least in some newer version of JAX, setting the flag will not be needed anymore.

**Reply by akuegel** [reply]
Short update: we have removed the legacy emitters and the flag in OpenXLA. So with the next JAX release, you won't need the flag anymore.
